File: api/routers/verification.py
import csv, io
from datetime import datetime
from typing import List, Optional
import httpx
from fastapi import APIRouter, BackgroundTasks, File, HTTPException, Query, UploadFile
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from ..services import zerobounce as zb
from ..services.job_store import (
    append_result, create_job, get_all_results, get_job, list_jobs, update_job
)

router = APIRouter(prefix="/verify", tags=["Verification"])

class VerifyRequest(BaseModel):
    email: str

class VerifyResult(BaseModel):
    email:  str
    status: str
    score:  int
    reason: Optional[str] = None
    checks: dict = {}

@router.post("/email", response_model=VerifyResult)
async def verify_single(req: VerifyRequest):
    async with httpx.AsyncClient() as client:
        try:
            result = await zb.verify(req.email, client)
            return VerifyResult(**result)
        except Exception as exc:
            raise HTTPException(502, f"Verification failed: {exc}")

@router.post("/bulk")
async def verify_bulk(
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    concurrency: int = Query(default=5, ge=1, le=20),
):
    if not file.filename or not file.filename.endswith(".csv"):
        raise HTTPException(400, "Only .csv files accepted.")
    raw  = await file.read()
    text = raw.decode("utf-8-sig") if raw[:3] == b'\xef\xbb\xbf' else raw.decode("latin-1")
    reader = csv.DictReader(io.StringIO(text))
    emails = [row.get("email","").strip() for row in reader if row.get("email","").strip()]
    if not emails:
        raise HTTPException(400, "No emails found in CSV.")
    job_id = await create_job(len(emails), job_type="verification")
    background_tasks.add_task(_run_bulk, job_id, emails, concurrency)
    return {"job_id": job_id, "total": len(emails), "status": "queued"}

@router.get("/jobs/{job_id}")
async def verify_job_status(job_id: str):
    job = await get_job(job_id)
    if not job: raise HTTPException(404, "Job not found.")
    return job

@router.get("/jobs/{job_id}/download")
async def download_verification(job_id: str):
    job = await get_job(job_id)
    if not job: raise HTTPException(404, "Job not found.")
    if job["status"] != "completed": raise HTTPException(409, "Job not complete.")
    results    = await get_all_results(job_id)
    fieldnames = ["email","status","score","reason","checks"]
    def _gen():
        buf = io.StringIO()
        w   = csv.DictWriter(buf, fieldnames=fieldnames, extrasaction="ignore")
        w.writeheader(); yield buf.getvalue()
        for row in results:
            buf = io.StringIO()
            w.writerow(row); yield buf.getvalue()
    return StreamingResponse(_gen(), media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename=verified_{job_id}.csv"})

async def _run_bulk(job_id: str, emails: List[str], concurrency: int):
    import asyncio
    await update_job(job_id, status="running", started_at=datetime.utcnow().isoformat())
    sem = asyncio.Semaphore(concurrency)
    async def _verify_one(email: str):
        async with sem:
            async with httpx.AsyncClient() as client:
                try:
                    result = await zb.verify(email, client)
                    await append_result(job_id, result)
                    await update_job(job_id,
                        processed_rows=1, successful_rows=1,
                        status="running")
                except Exception as e:
                    await append_result(job_id, {"email": email, "status": "unknown", "score": 0, "reason": str(e)})
                    await update_job(job_id, processed_rows=1, failed_rows=1)
    await asyncio.gather(*[_verify_one(e) for e in emails])
    await update_job(job_id, status="completed", completed_at=datetime.utcnow().isoformat())

File: api/routers/verification.js
import { Router } from "express";
import multer from "multer";
import { parse } from "csv-parse/sync";
import { stringify } from "csv-stringify/sync";
import { verify } from "../services/zerobounce.js";
import { createJob, getJob, updateJob, appendResult, getAllResults } from "../services/jobStore.js";

export const verificationRouter = Router();
const upload = multer({ storage: multer.memoryStorage() });

verificationRouter.post("/verify/email", async (req, res) => {
  const { email } = req.body;
  if (!email) return res.status(422).json({ error: "email is required." });
  try {
    const result = await verify(email);
    res.json(result);
  } catch (e) {
    res.status(502).json({ error: `Verification failed: ${e.message}` });
  }
});

verificationRouter.post("/verify/bulk", upload.single("file"), async (req, res) => {
  if (!req.file) return res.status(400).json({ error: "No file uploaded." });
  const concurrency = Math.min(Number(req.query.concurrency) || 5, 20);
  let records;
  try {
    records = parse(req.file.buffer.toString("utf-8"), { columns: true, skip_empty_lines: true, trim: true });
  } catch (e) {
    return res.status(400).json({ error: `CSV parse error: ${e.message}` });
  }
  const emails = records.map(r => r.email || r.Email || "").filter(Boolean);
  if (!emails.length) return res.status(400).json({ error: "No emails found in CSV." });
  const jobId = await createJob(emails.length, "verification");
  res.json({ job_id: jobId, total: emails.length, status: "queued" });
  setImmediate(() => _runBulkVerify(jobId, emails, concurrency));
});

async function _runBulkVerify(jobId, emails, concurrency) {
  await updateJob(jobId, { status: "running", started_at: new Date().toISOString() });
  const queue   = [...emails];
  const workers = Array.from({ length: concurrency }, async () => {
    while (queue.length) {
      const email = queue.shift();
      try {
        const result = await verify(email);
        await appendResult(jobId, result);
        await updateJob(jobId, { processed_rows: 1, successful_rows: 1 });
      } catch (e) {
        await appendResult(jobId, { email, status: "unknown", score: 0, reason: e.message });
        await updateJob(jobId, { processed_rows: 1, failed_rows: 1 });
      }
    }
  });
  await Promise.all(workers);
  await updateJob(jobId, { status: "completed", completed_at: new Date().toISOString() });
}

verificationRouter.get("/verify/jobs/:jobId", async (req, res) => {
  const job = await getJob(req.params.jobId);
  if (!job) return res.status(404).json({ error: "Job not found." });
  res.json(job);
});

verificationRouter.get("/verify/jobs/:jobId/download", async (req, res) => {
  const job = await getJob(req.params.jobId);
  if (!job) return res.status(404).json({ error: "Job not found." });
  if (job.status !== "completed") return res.status(409).json({ error: "Not complete." });
  const results = await getAllResults(req.params.jobId);
  const csv = stringify(results, { header: true, columns: ["email","status","score","reason"] });
  res.setHeader("Content-Type", "text/csv");
  res.setHeader("Content-Disposition", `attachment; filename=verified_${req.params.jobId}.csv`);
  res.send(csv);
});
